File: ImageSliderWithIndicatorExample/app/src/main/java/com/example/velmurugan/imagesliderwithindicatorexample/RecyclerviewAdapter.java
package com.example.velmurugan.imagesliderwithindicatorexample;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.List;


public class RecyclerviewAdapter  extends RecyclerView.Adapter<RecyclerviewAdapter.MyviewHolder> {

    private List<Product> products;
    private Context context;

    RecyclerviewAdapter(Context context,List<Product> products){
        this.context = context;
        this.products = products;
    }

    public void setProducts(List<Product> products) {
        this.products = products;
        notifyDataSetChanged();
    }

    @Override
    public RecyclerviewAdapter.MyviewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.recyclarview_adapter,parent,false);
        return new MyviewHolder(view);
    }

    @Override
    public void onBindViewHolder(RecyclerviewAdapter.MyviewHolder holder, int position) {

        holder.tvOffer.setText(products.get(position).getpOffer());
        holder.tvTittle.setText(products.get(position).getpName());
        Picasso.with(context).load(products.get(position).getpImage()).into(holder.imgTitle);
    }

    @Override
    public int getItemCount() {
        if(products != null){
            return products.size();
        }
        return 0;
    }

    public class MyviewHolder extends RecyclerView.ViewHolder {

        TextView tvOffer,tvTittle;
        ImageView imgTitle;

        public MyviewHolder(View itemView) {
            super(itemView);
            tvOffer = (TextView)itemView.findViewById(R.id.tvOffer);
            tvTittle = (TextView)itemView.findViewById(R.id.tvTitle);
            imgTitle = (ImageView)itemView.findViewById(R.id.imgTitle);
        }
    }

}

File: CardviewExample/app/src/main/java/com/example/velmurugan/cardviewexample/RecyclerViewAdapter.kt
package com.example.velmurugan.cardviewexample

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.example.velmurugan.cardviewexample.RecyclerViewAdapter.MyViewHolder

class RecyclerViewAdapter constructor(private val movieList: List<Movie>) :
    RecyclerView.Adapter<MyViewHolder>() {
    private var clickListener: ClickListener<Movie>? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.recyclerview_adapter_layout, parent, false)
        return MyViewHolder(view)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        val movie = movieList[position]
        holder.title.text = movie.title
        holder.image.setBackgroundResource(movie.image)
        holder.cardView.setOnClickListener { clickListener!!.onItemClick(movie) }
    }

    override fun getItemCount(): Int {
        return movieList.size
    }

    fun setOnItemClickListener(movieClickListener: ClickListener<Movie>?) {
        clickListener = movieClickListener
    }

    inner class MyViewHolder(itemView: View) : ViewHolder(itemView) {
        val title: TextView = itemView.findViewById(R.id.title)
        val image: ImageView = itemView.findViewById(R.id.image)
        val cardView: CardView = itemView.findViewById(R.id.carView)

    }
}

interface ClickListener<T> {
    fun onItemClick(data: T)
}